internal/resolution/manifest/npm.go

```diff
@@ -79,50 +79,74 @@ func (rw NpmManifestIO) Read(f lockfile.DepFile) (Manifest, error) {
 		}
 		manif.LocalManifests = append(manif.LocalManifests, m)
 		workspaceNames[m.Root.Name] = struct{}{}
-		// TODO: Workspaces can potentially have name collisions with real packages
-		// The OverrideClient will always pick the local manifest, even in indirect dependencies.
-		// These workspace names probably need some unique identifier assigned to them...
-		// (what does npm actually do in this situation?)
+	}
+
+	isWorkspace := func(req resolve.RequirementVersion) bool {
+		if req.Type.HasAttr(dep.KnownAs) {
+			// "alias": "npm:pkg@*" seems to always take the real 'pkg',
+			// even if there's a workspace with the same name.
+			return false
+		}
+		_, ok := workspaceNames[req.Name]
+
+		return ok
 	}
 
 	workspaceReqVers := make(map[resolve.PackageKey]resolve.RequirementVersion)
 
+	// empirically, the dev version takes precedence over optional, which takes precedence over regular, if they conflict.
 	for pkg, ver := range packagejson.Dependencies {
-		dep := rw.makeNPMReqVer(pkg, ver)
-		if _, ok := workspaceNames[pkg]; ok {
+		req := rw.makeNPMReqVer(pkg, ver)
+		if isWorkspace(req) {
 			// workspaces seem to always be evaluated separately
-			workspaceReqVers[dep.PackageKey] = dep
+			workspaceReqVers[req.PackageKey] = req
 			continue
 		}
-		manif.Requirements = append(manif.Requirements, dep)
+		manif.Requirements = append(manif.Requirements, req)
+	}
+
+	for pkg, ver := range packagejson.OptionalDependencies {
+		req := rw.makeNPMReqVer(pkg, ver)
+		req.Type.AddAttr(dep.Opt, "")
+		if isWorkspace(req) {
+			// workspaces seem to always be evaluated separately
+			workspaceReqVers[req.PackageKey] = req
+			continue
+		}
+		if idx := slices.IndexFunc(manif.Requirements, func(imp resolve.RequirementVersion) bool {
+			return imp.PackageKey == req.PackageKey
+		}); idx != -1 {
```

Comment:
This is a little hard to read imo, given the complexity of the previous if condition.

It's probably batter to just break this out into its own if.